_posts/2023-03-22-dd_00005_half_adder.md

```diff
@@ -0,0 +1,112 @@
+as [this](https://www.youtube.com/watch?v=LpuPe81bc2w) one on the basics of
+binary numbers, followed by [this](https://www.youtube.com/watch?v=C5EkxfNEMjE)
+one on doing addition in the binary system. We might also cover binary numbers
+in a future post in our [Getting Started](/categories/getting_started) category,
+so stay tuned for that. Ok, you know what binary numbers are and how to add
+them? Perfect! Then let's see how we can build a circuit that can do binary
+addition.
+
+To keep things simple, let's first look at adding two **1-bit** numbers. Recall,
+that, if both bits are 0 then the sum is 0, obviously. If one of the bit is 1
+and the other is 0, then the sum is a 1. And finally, if both bits are 1, then
+the sum will be 0 and we get a _carry_ of 1. We can visualize this in a table
+form (calling the two bits “A” and “B”, respectively):
+
+| A   | B   | Sum | Carry |
+| --- | --- | --- | ----- |
+| 0   | 0   | 0   | 0     |
+| 0   | 1   | 1   | 0     |
+| 1   | 0   | 1   | 0     |
+| 1   | 1   | 0   | 1     |
+
+Do you notice something? Right, this looks like a logic gate **truth table**!
+And that's precisely how we can come up with a logic circuit for adding two
+1-bit numbers. This circuit has two inputs, the two numbers A and B, and two
+outputs, the sum and the carry. The corresponding circuit looks as follows:
+
+<figure class="img-center" style="width:50%">
+  <img src="/assets/images/dd-00005-half_adder_circuit.png" alt="Half Adder" title="Half Adder">
+  <!-- <figcaption class="figure-caption text-center">Microchips</figcaption> -->
+</figure>
+
+This circuit looks pretty simple, just one **XOR** gate whose output is the sum
+bit and an **AND** gate where the carry comes out. If you look at the truth
+table above, and remember how the truth tables of the AND and XOR gate look
+like, you can easily see why we need these two gates. This circuit is called
+**Half-Adder**. To not have to redraw all the gates when we are drawing many
+half-adders, engineers came up with a new symbol to depict a half-adder:
+
+<figure class="img-center" style="width:50%">
+  <img src="/assets/images/dd-00005-half_adder_symbol.png" alt="Half Adder Symbol" title="Half Adder Symbol">
+  <!-- <figcaption class="figure-caption text-center">Microchips</figcaption> -->
+</figure>
+
+<div class='note-normal'>
+This is yet another example for <i>abstraction</i>: We simplify things by hiding the
+lower-level details of a particular system. In our case, the low-level details
+were the AND and XOR gates of the half-adder. And remember, those gates
+themselves again were an abstraction of the underlying transisitor circuits!
+</div>
+
+You might think that now extending it to adding larger than 1-bit numbers is as
+simple as taking a bunch of half-adders for each digit, right? Well, almost. We
+need to extend our circuit a bit. Why? Because we also need to consider the
+_carry_ bit coming from the _previous_ digits (except the first digits). The
```

Comment:
```suggestion
_carry_ bit coming from the _previous_ digits (except for the first digits). The
```

Reply:
Otherwise, it sounds a bit like we ignore the carry coming from the first digits